storage: Add support for TargetBytes for EndTxn

KaiSun314 opened the pull request:
Fixes: #77228

Intent resolution batches are sequenced on raft and each batch can
consist of 100-200 intents. If an intent key or even value in some cases
are large, it is possible that resolving all intents in the batch would
result in a raft command size exceeding the max raft command size
kv.raft.command.max_size.

To address this, we add support for TargetBytes in resolve intent and
resolve intent range commands, allowing us to stop resolving intents in
the batch as soon as we exceed the TargetBytes max bytes limit.

This PR adds byte pagination for synchronous intent resolution (i.e.
EndTxn / End Transaction).

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/95141)
<!-- Reviewable:end -->

KaiSun314 commented:
**Please only review the fourth commit `storage: Add byte pagination to synchronous intent resolution / EndTxn`**. The first commit `storage: Add support for TargetBytes for resolve intent + range cmd` is https://github.com/cockroachdb/cockroach/pull/94814, the second commit `storage: Add MVCCPagination to allow general key and byte pagination` and third commit `storage: Push pagination for EndTxn into the MVCC layer` are #94939. Thank you so much in advance!

KaiSun31415 committed:
storage: Add byte pagination to synchronous intent resolution / EndTxn

Informs: #77228

Intent resolution batches are sequenced on raft and each batch can
consist of 100-200 intents. If an intent key or even value in some cases
are large, it is possible that resolving all intents in the batch would
result in a raft command size exceeding the max raft command size
kv.raft.command.max_size.

To address this, we add support for TargetBytes in resolve intent and
resolve intent range commands, allowing us to stop resolving intents in
the batch as soon as we exceed the TargetBytes max bytes limit.

This PR adds byte pagination for synchronous intent resolution (i.e. the
EndTxn / End Transaction command).

Release note: None

nvb reviewed (commented):
I just looked at the third commit here, as the first two are in https://github.com/cockroachdb/cockroach/pull/94939. It's a very nice change!

Is there more to do for https://github.com/cockroachdb/cockroach/issues/77228 once your 3 in-progress PRs merge?

> Reviewed 5 of 5 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95141#-:-NOwln0fEprtR8Kj1tnM:bpchp5x)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @KaiSun314)

___
*[`pkg/kv/kvserver/batcheval/cmd_end_transaction.go` line 535 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95141#-NOwkczU5ZHW0C-0Xt25:-NOwkczU5ZHW0C-0Xt26:b-xap4t3) ([raw file](https://github.com/cockroachdb/cockroach/blob/01a018b020b68b1182af07256e259185576ca81e/pkg/kv/kvserver/batcheval/cmd_end_transaction.go#L535)):*
> ```Go
> 
> // resolveLocalLocksWithPagination is resolveLocalLocks but with a max key
> // limit.
> ```

"with a max key and target bytes limit"
___
*[`pkg/kv/kvserver/batcheval/cmd_end_transaction.go` line 587 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95141#-NOwk5vmEi37Todf_u9D:-NOwk5vmEi37Todf_u9E:bbu988c) ([raw file](https://github.com/cockroachdb/cockroach/blob/01a018b020b68b1182af07256e259185576ca81e/pkg/kv/kvserver/batcheval/cmd_end_transaction.go#L587)):*
> ```Go
> 					numKeys = 1
> 				}
> 				if resumeSpan != nil {
> ```

Do you mind leaving a comment around here that `!ok && resumeSpan == nil` is a valid condition that means that no intent was found. It's tempting to think that some of this branching could be simplified.


<!-- Sent from Reviewable.io -->

KaiSun314 reviewed (commented):
Thank you so much for the review Nathan!

I believe once these 2 synchronous intent resolution PRs merge (I saw the 1 remaining asynchronous intent resolution PR has  been successfully merged, thank you for merging!), #77228 should be complete :)

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95141#-:-NP0P0492PlP2Yy-UhvG:b-pxulq4)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten)



<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
@KaiSun314 please rebase this on master at your convenience, then we can get it in. Thanks!

KaiSun314 commented:
Ah sorry about that Nathan, I have rebased on master now. Thank you so much for the review and for getting this in!

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

bors r=nvanbenschoten

> Reviewed 2 of 5 files at r6, 2 of 3 files at r9, 4 of 4 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95141#-:-NPTLSffCRZPGSVlIWdF:b-qdhl0k)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @KaiSun314)



<!-- Sent from Reviewable.io -->
